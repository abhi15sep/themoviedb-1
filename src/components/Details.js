import React, { Component } from 'react';
import Loading from './Loading';
import Error from './Error';
import Image from './Image';
import CONSTANTS from '../constants';
import { Grid, Row, Col } from 'react-bootstrap';
import api from '../api';
import './Details.css';

const { SIZE } = CONSTANTS.IMAGE;

class Details extends Component {

  constructor(props) {
    super(props);
    this.state = {
      loading: true,
      error: false,
      movie: {},
    };
  }

  componentDidMount() {
    this.getMovie();
  }

  getMovie() {
    api.getMovie(this.props.match.params.id)
    .then(response => {
      if (response.ok) {
        return response.json();
      } else {
        throw Error(response.statusText);
      }
    })
    .then(data => {
      this.setState({
        loading: false,
        movie: data,
      });
    })
    .catch(error => {
      this.setState({
        error: true
      });
    });
  }

  getGenres () {
    const { genres } = this.state.movie;

    if (genres.length > 0) {
      const list = genres.map(genre => genre.name).join(', ');
      
      return (
        < div className="Details-genres">
          <h2>Genres</h2>
          <p>{list}</p>
        </div>
      );
    }

    return null;
  }

  getHomepage () {
    const { homepage, title } = this.state.movie;

    if (homepage) {
      return (
        <div className="Details-homepage">
          <h2>Homepage</h2>
          <p><a href={homepage} alt={title} title={title}>{homepage}</a></p>
        </div>
      );
    }

    return null;
  }

  renderMovie() {
    const {
      title,
      poster_path,
      overview,
      release_date,
      vote_average,
    } = this.state.movie;

    const releaseYear = new Date(release_date).getFullYear();
    
    return (
      <Grid className="Details-main">
      <Row>
        <Col xs={12} md={4}>
          <Image size={SIZE.LARGE} path={poster_path} title={title}/>
        </Col>

        <Col xs={12} md={8} className="Details-body">
          <h1>
            <span className={`label label-default`}>{vote_average}</span>
            &nbsp;
            {title}
            &nbsp;
            <small>({releaseYear})</small>
          </h1>

          <div className="Details-overview">
            <p>{overview}</p>
          </div>

           {this.getGenres()}

          {this.getHomepage()}
        </Col>
      </Row>
    </Grid>
    );
  }

  render () {
    const {
      error, 
      loading,
    } = this.state;

    if (error) {
      return <Error/>;
    } else if (loading) {
      return <Loading/>;
    }

    return this.renderMovie();
  }
}

export default Details;
