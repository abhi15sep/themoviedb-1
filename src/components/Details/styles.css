.Details-main {
    background-color: #fff;
    box-shadow: 0 2px 8px rgba(0,0,0,0.1);
    margin-top: 2em;
    margin-bottom: 2em;
    position: relative;
}

.Details-title {
    margin-top: 1em;
    padding-left: 0 !important;
}

.Details-body {
    text-align: left;
    padding-left: 50px;
    padding-right: 50px;
}

.Details-body a {
    text-decoration: underline;
}

.Details-overview p {
    text-align: justify;
    margin-top: 2em;
    margin-right: 2em;
    font-size: 1.3em;
}

.Details-body h2 {
    font-size: 1.3em;
    font-weight: 500;
}

.Details-genre {
    margin-right: .5em;
}
