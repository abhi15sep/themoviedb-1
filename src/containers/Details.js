import React, { Component } from 'react';
import { Grid, Row, Col } from 'react-bootstrap';
import PropTypes from 'prop-types';
import Loading from '../components/Loading';
import Error from '../components/Error';
import Image from '../components/Image';
import CONSTANTS from '../constants';
import api from '../api/';
import './Details.css';

const { SIZE } = CONSTANTS.IMAGE;

class Details extends Component {
  constructor(props) {
    super(props);
    this.state = {
      loading: true,
      error: false,
      movie: {},
    };
  }

  componentDidMount() {
    this.isAlreadyMounted = true;
    this.getMovie();
  }

  componentWillUnmount() {
    this.isAlreadyMounted = false;
  }

  getMovie() {
    api.getMovie(this.props.match.params.id)
      .then((data) => {
        if (this.isAlreadyMounted) {
          this.setState({
            loading: false,
            movie: data,
          });
        }
      })
      .catch(() => {
        if (this.isAlreadyMounted) {
          this.setState({
            error: true,
          });
        }
      });
  }

  getGenres() {
    const { genres } = this.state.movie;

    if (genres && genres.length > 0) {
      const list = genres.map(genre => genre.name).join(', ');

      return (
        <div className="Details-genres">
          <h2>Genres</h2>
          <p>{list}</p>
        </div>
      );
    }

    return null;
  }

  getHomepage() {
    const { homepage, title } = this.state.movie;

    if (homepage) {
      return (
        <div className="Details-homepage">
          <h2>Homepage</h2>
          <p><a href={homepage} alt={title} title={title}>{homepage}</a></p>
        </div>
      );
    }

    return null;
  }

  renderMovie() {
    const {
      title,
      overview,
      poster_path: posterPath,
      release_date: releaseDate,
      vote_average: voteAverage,
    } = this.state.movie;

    const releaseYear = new Date(releaseDate).getFullYear();

    return (
      <Grid className="Details-main">
        <Row>
          <Col xs={12} md={4}>
            <Image size={SIZE.LARGE} path={posterPath} title={title} />
          </Col>

          <Col xs={12} md={8} className="Details-body">
            <h1>
              <span className="label label-default">{voteAverage}</span>
              &nbsp;
              {title}
              &nbsp;
              <small>({releaseYear})</small>
            </h1>

            <div className="Details-overview">
              <p>{overview}</p>
            </div>

            {this.getGenres()}

            {this.getHomepage()}
          </Col>
        </Row>
      </Grid>
    );
  }

  render() {
    const {
      error,
      loading,
    } = this.state;

    if (error) {
      return <Error />;
    } else if (loading) {
      return <Loading />;
    }

    return this.renderMovie();
  }
}

Details.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
};


export default Details;
