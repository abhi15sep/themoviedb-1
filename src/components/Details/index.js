import React from 'react';
import {
  Container,
  Row,
  Col,
  Badge,
} from 'react-bootstrap';
import Image from '../Image';
import CONSTANTS from '../../utils/constants';
import getLabelInfo from '../../utils/getLabelInfo';
import Types from '../../utils/types';
import './styles.css';

const { SIZE } = CONSTANTS.IMAGE;

function renderImage(title, posterPath) {
  return <Image size={SIZE.LARGE} path={posterPath} title={title} />;
}

function renderTitle(title, releaseDate, voteAverage) {
  const voteAverageNormalized = Math.floor(voteAverage);
  const labelInfo = getLabelInfo(voteAverageNormalized);
  const releaseYear = new Date(releaseDate).getFullYear();

  return (
    <Container className="Details-title">
      <Row>
        <Col xs={2}>
          <h1>
            <Badge variant={labelInfo}>
              {voteAverage}
            </Badge>
          </h1>
        </Col>
        <Col>
          <h1>
            {title}
            &nbsp;
            <small>
              (
              {releaseYear}
              )
            </small>
          </h1>
        </Col>
      </Row>
    </Container>
  );
}

function renderOverview(overview) {
  return (
    <div className="Details-overview">
      <p>{overview}</p>
    </div>
  );
}

function renderGenres(genres) {
  if (genres && genres.length > 0) {
    const list = genres.map((genre) => genre.name).join(', ');

    return (
      <div className="Details-genres">
        <h2>Genres</h2>
        <p>{list}</p>
      </div>
    );
  }

  return null;
}

function renderHomepage(homepage, title) {
  if (homepage) {
    return (
      <div className="Details-homepage">
        <h2>Homepage</h2>
        <p><a href={homepage} alt={title} title={title}>{homepage}</a></p>
      </div>
    );
  }

  return null;
}

function Details({ movie }) {
  const {
    overview,
    genres,
    homepage,
    title,
    poster_path: posterPath,
    release_date: releaseDate,
    vote_average: voteAverage,
  } = movie;

  return (
    <Container className="Details-main">
      <Row>
        <Col xs={12} md={4} className="Details-image">
          {renderImage(title, posterPath)}
        </Col>

        <Col xs={12} md={8} className="Details-body">
          {renderTitle(title, releaseDate, voteAverage)}
          {renderOverview(overview)}
          {renderGenres(genres)}
          {renderHomepage(homepage, title)}
        </Col>
      </Row>
    </Container>
  );
}

Details.propTypes = {
  movie: Types.movie.isRequired,
};


export default Details;
