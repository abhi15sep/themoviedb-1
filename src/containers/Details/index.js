import React, { Component } from 'react';
import { Grid, Row, Col } from 'react-bootstrap';
import PropTypes from 'prop-types';
import Search from '../Search';
import Loading from '../../components/Loading';
import Error from '../../components/Error';
import Image from '../../components/Image';
import CONSTANTS from '../../utils/constants';
import api from '../../api/';
import './styles.css';

const { SIZE } = CONSTANTS.IMAGE;

class Details extends Component {
  constructor(props) {
    super(props);
    this.state = {
      id: this.props.match.params.id,
      loading: true,
      error: false,
      movie: {},
    };
  }

  componentDidMount() {
    this.isAlreadyMounted = true;
    this.getMovie();
  }

  componentDidUpdate(prevProps, prevState) {
    const { id } = this.props.match.params;

    if (prevState.id !== id) {
      this.getMovie();
    }
  }

  componentWillUnmount() {
    this.isAlreadyMounted = false;
  }

  getMovie() {
    const { id } = this.props.match.params;

    api.getMovie(id)
      .then((data) => {
        if (this.isAlreadyMounted) {
          this.setState({
            id,
            loading: false,
            movie: data,
          });
        }
      })
      .catch(() => {
        if (this.isAlreadyMounted) {
          this.setState({
            error: true,
          });
        }
      });
  }

  getGenres() {
    const { genres } = this.state.movie;

    if (genres && genres.length > 0) {
      const list = genres.map(genre => genre.name).join(', ');

      return (
        <div className="Details-genres">
          <h2>Genres</h2>
          <p>{list}</p>
        </div>
      );
    }

    return null;
  }

  getHomepage() {
    const { homepage, title } = this.state.movie;

    if (homepage) {
      return (
        <div className="Details-homepage">
          <h2>Homepage</h2>
          <p><a href={homepage} alt={title} title={title}>{homepage}</a></p>
        </div>
      );
    }

    return null;
  }

  renderMovie() {
    const {
      title,
      overview,
      poster_path: posterPath,
      release_date: releaseDate,
      vote_average: voteAverage,
    } = this.state.movie;

    const releaseYear = new Date(releaseDate).getFullYear();

    return (
      <div className="Details-wrapper">
        <Search />
        <Grid className="Details-main">
          <Row>
            <Col xs={12} md={4}>
              <Image size={SIZE.LARGE} path={posterPath} title={title} />
            </Col>

            <Col xs={12} md={8} className="Details-body">
              <h1>
                <span className="label label-default">{voteAverage}</span>
                &nbsp;
                {title}
                &nbsp;
                <small>({releaseYear})</small>
              </h1>

              <div className="Details-overview">
                <p>{overview}</p>
              </div>

              {this.getGenres()}

              {this.getHomepage()}
            </Col>
          </Row>
        </Grid>
      </div>
    );
  }

  render() {
    const {
      error,
      loading,
    } = this.state;

    if (error) {
      return <Error />;
    } else if (loading) {
      return <Loading />;
    }

    return this.renderMovie();
  }
}

Details.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
};


export default Details;
