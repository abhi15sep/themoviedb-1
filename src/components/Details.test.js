import React from 'react';
import ReactDOM from 'react-dom';
import Details from './Details';

const props = {
  match: {
    params: {
      id: '1234',
    },
  },
};

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<Details {...props} />, div);
  ReactDOM.unmountComponentAtNode(div);
});
