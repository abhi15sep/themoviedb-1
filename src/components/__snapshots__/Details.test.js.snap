// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="Loading-main"
>
  <img
    alt="spinner"
    className="Loading-spinner"
    src="spinner.svg"
  />
  <p>
    Loading...
  </p>
</div>
`;
